bugsnag-android-core/src/main/java/com/bugsnag/android/DeliveryHeaders.kt

```diff
@@ -12,19 +16,44 @@ internal const val HEADER_INTERNAL_ERROR = "Bugsnag-Internal-Error"
  *
  * @return the HTTP headers
  */
-internal fun errorApiHeaders(apiKey: String) = mapOf(
-    HEADER_API_PAYLOAD_VERSION to "4.0",
-    HEADER_API_KEY to apiKey,
-    HEADER_BUGSNAG_SENT_AT to DateUtils.toIso8601(Date())
-)
+internal fun errorApiHeaders(payload: EventPayload): Map<String, String?> {
+    return mapOf(
+        HEADER_API_PAYLOAD_VERSION to "4.0",
+        HEADER_API_KEY to payload.apiKey,
+        HEADER_BUGSNAG_SENT_AT to DateUtils.toIso8601(Date()),
+        HEADER_BUGSNAG_INTEGRITY to computeSha1Digest(payload)
+    )
+}
 
 /**
  * Supplies the headers which must be used in any request sent to the Session Tracking API.
  *
  * @return the HTTP headers
  */
-internal fun sessionApiHeaders(apiKey: String) = mapOf(
+internal fun sessionApiHeaders(apiKey: String): Map<String, String?> = mapOf(
     HEADER_API_PAYLOAD_VERSION to "1.0",
     HEADER_API_KEY to apiKey,
     HEADER_BUGSNAG_SENT_AT to DateUtils.toIso8601(Date())
 )
+
+internal fun computeSha1Digest(payload: EventPayload): String? {
+    runCatching {
+        val shaDigest = MessageDigest.getInstance("SHA-1")
+        val builder = StringBuilder("sha1 ")
+
+        // Pipe the object through a no-op output stream
+        DigestOutputStream(NullOutputStream(), shaDigest).use { stream ->
+            stream.bufferedWriter().use { writer ->
+                payload.toStream(JsonStream(writer))
```

Comment:
It looks like DefaultDelivery is already serialising to JSON. Can't we use that data instead of serialising a second time?

Reply:
As the request uses chunked transfer encoding to minimise memory footprint, the Java object/File is streamed directly into the HTTP request, meaning the payload size is not known ahead of time. I don't think it's possible to get around needing to serialize twice without sacrificing chunked transfer encoding - and this is what was chosen during the design phase.